src/certificates.rs

```diff
@@ -0,0 +1,228 @@
+            }
+            Err(CertLoadError::BadCert(_, e)) => {
+                return Err(CertLoadError::BadCert("fallback".to_string(), e))
+            }
+            Err(CertLoadError::MissingKey(_)) => {
+                return Err(CertLoadError::MissingKey("fallback".to_string()))
+            }
+            Err(CertLoadError::MissingCert(_)) => {
+                return Err(CertLoadError::MissingCert("fallback".to_string()))
+            }
+            // For the fallback keys there is no domain name to verify them
+            // against, so we can skip that step and only have to do it for the
+            // other keys below.
+            Ok(key) => certs.push((String::new(), key)),
+        }
+
+        for file in certs_dir
+            .read_dir()
+            .or(Err(CertLoadError::NoReadCertDir))?
+            .filter_map(Result::ok)
+            .filter(|x| x.path().is_dir())
+        {
+            let path = file.path();
+
+            // the filename should be the domain name
+            let filename = path
+                .file_name()
+                .and_then(OsStr::to_str)
+                .unwrap()
+                .to_string();
+
+            let dns_name = match DNSNameRef::try_from_ascii_str(&filename) {
+                Ok(name) => name,
+                Err(_) => return Err(CertLoadError::BadDomain(filename)),
+            };
+
+            let key = load_domain(certs_dir, filename.clone())?;
+            key.cross_check_end_entity_cert(Some(dns_name))
+                .or_else(|e| Err(CertLoadError::BadCert(filename.clone(), e.to_string())))?;
+
+            certs.push((filename, key));
+        }
+
+        certs.sort_unstable_by(|(a, _), (b, _)| {
+            // Try to match as many domain segments as possible. If one is a
+            // substring of the other, the `zip` will only compare the smaller
+            // length of either a or b and the for loop will not decide.
+            for (a_part, b_part) in a.split('.').rev().zip(b.split('.').rev()) {
+                if a_part != b_part {
+                    // What we sort by here is not really important, but `str`
+                    // already implements Ord, making it easier for us.
+                    return a_part.cmp(b_part);
+                }
+            }
+            // Sort longer domains first.
+            a.len().cmp(&b.len()).reverse()
+        });
+
+        Ok(Self { certs })
+    }
+}
+
+impl ResolvesServerCert for CertStore {
+    fn resolve(&self, client_hello: rustls::ClientHello<'_>) -> Option<CertifiedKey> {
+        if let Some(name) = client_hello.server_name() {
+            let name: &str = name.into();
+            // The certificate list is sorted so the longest match will always
+            // appear first. We have to find the first that is either this
+            // domain or a parent domain of the current one.
+            self.certs
+                .iter()
+                .find(|(s, _)| name.ends_with(s))
```

Comment:
I think we might be able to use `binary_search_by` here, but I am not sure if this would always find the longest possible match.

Reply:
It is not possible to use `binary_search_by` here, because it would not find something in all cases where it is necessary to find something, for example: `d.c.b.a`, `c.b.a`, `e.b.a` and `b.a` are present, searching for `b.a` yields an `Err` and the index will point to `d.c.b.a`.